[#2286]: Underscore could use a _.propertyResult

monkpow opened the pull request:
I chose the name _.method per the above suggestion, but am happy to adjust if this does not match the project guidelines.

megawac commented on underscore.js:
```diff
@@ -1370,6 +1370,24 @@
+  // referenced object.
+  _.method = function(value) {
+    var isFunction = this.isFunction,
+        rest = this.rest,
```
use `restArgs` instead

megawac commented on underscore.js:
```diff
@@ -1370,6 +1370,24 @@
+  // If value is a string, function assumes (and verifies) this string is a method name on the
+  // referenced object.
+  _.method = function(value) {
+    var isFunction = this.isFunction,
```
don't use `this` to reference underscore functions. Use `_`

megawac commented on underscore.js:
```diff
@@ -1370,6 +1370,24 @@
+        func;
+
+    return function(obj) {
+      if (obj == null) { return; }
```
is this not valid when a function is provided?

monkpow replied:
This method responds to both a string literal (a method name) or a function literal.   This check is to sort out if the value passed is a function literal, so that later we know what to call apply on.

megawac commented on underscore.js:
```diff
@@ -1370,6 +1370,24 @@
+
+    return function(obj) {
+      if (obj == null) { return; }
+      func = isFunction(value) ? value : obj[value];
```
could check if `value` is a function outside of this function

monkpow commented:
@megawac Thanks for the quality feedback.  Updated diff addresses:
- use restArgs
- don't use `this` keyword, use direct reference
- Clarified comment and token in response to Question: "is this not valid when a function is provided?".  Added narrative above in this thread.
- check if value is a function in the outer function.

megawac commented on underscore.js:
```diff
@@ -1370,6 +1370,24 @@
+
+    return restArgs(function(obj, args) {
+      if (obj == null) { return; }
+      func = isFunctionLiteral ? value : obj[value];
```
define `func` here rather than outer scope

monkpow replied:
Thanks, updated diff with this change.

megawac commented on underscore.js:
```diff
@@ -1370,6 +1370,24 @@
+        func;
+
+    return restArgs(function(obj, args) {
+      if (obj == null) { return; }
```
any rational why should func not be called if it is a function literal and obj is null?

monkpow replied:
My current rational is that this invocation pattern doesn't make much sense to me.  e.g.,

```
_().method(function() { /* function body *. })
_.method(null, function () { /* function body */ })
```

since both would be more easily written as a plain function literal, and this invocation pattern reads to me as 'call the provided method on undefined'.

That said, if allowing an undefined object in this position is more in keeping with underscore's style, I'm happy to update so we can call _.method on undefined objects.

jdalton commented:
fyi [lodash/method](https://github.com/lodash/lodash/blob/4.11.2-npm/method.js).
